### ISSUE TITLE
Log schemas

### ISSUE THREAD

**Issue by vector-vic** [vectordotdev/vector#3910]
A common need for Vector users is the ability to map data according to different schemas. This is a key requirement for Vector since it aims to be schema, standard, and vendor-neutral. In order to deliver on this claim, Vector must not only support a variety of schemas independently, but it must also assist in the interchange between them.

# Use Cases

## Transitioning to Vector

Schemas create very heavy lock-in. This is because most downstream systems depend on this schema. To name a few:

1. Alerts.

2. Graphs and dashboards.

3. Storages.

4. Humans.

Changing a schema can break all of these things which usually is not acceptable. To prevent this, Vector must adopt their current schema in a way that downstream dependencies do not notice.

## Transitioning Vendors

The use case above illustrates the need for Vector to support a single schema at a time, but there are cases where a user would need to support multiple. For example, when transitioning vendors. Vector must not only support the "read" schema but also transform the data to an entirely new "write" schema.

For example, if a user is transitioning from Splunk to Elasticsearch, Vector must ingest the data under the [Splunk Common Information Model](https://docs.splunk.com/Documentation/CIM/4.17.0/User/Overview) and transform it to the [Elastic Common Schema](https://www.elastic.co/guide/en/ecs/current/index.html).

## Automatic Dashboards, Alerts, & Insights

A benefit of using a vendor's agent is that it'll unlock automatic dashboards, alerts, and other features. This not only saves a considerable amount of time and effort, but you can effectively delegate the management of these things to your chosen vendor. For example, I assume that DataDog, and their community, continually improve their dashboards. In this case, it's very important that Vector can transparently adopt the DataDog schema so that DataDog Vector users can receive the same benefit. It also alleviates us from having to maintain these entities as well. 

# Schemas

1. [Elastic Common Schema](https://www.elastic.co/guide/en/ecs/current/index.html)

2. [Splunk Common Information Model (CIM)](https://docs.splunk.com/Documentation/CIM/4.17.0/User/Overview)

3. [OpenTelemetry Log Data Model](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/logs/data-model.md)

4. [GELF](https://docs.graylog.org/en/3.2/pages/gelf.html#gelf-payload-specification)

5. [DataDog's reserved log attributes](https://docs.datadoghq.com/logs/log_collection/?tab=http#reserved-attributes)

6. ...and more

# Proposal

In short, I'm proposing that we attach the known schema to each Vector event during ingestion. This would allow us to lookup fields and map them across schemas. There are a lot of little details to discuss which we can cover in an RFC. To name a few:

1. How would Vector detect the schema?

2. Should Vector strictly enforce the schema? Ex: Not allowing users to add fields that would violate the schema.

3. Should Vector reject data at the source-level that does not conform to the chosen schema?

4. Should Vector adopt a default schema? Ex: OpenTelemetry.

5. What happens when the user has a custom schema that we know nothing about? Ex: require them to manually map data when necessary.

**Comment by vector-vic** [comment]
Link to feature: https://timber.productboard.com/feature-board/planning/features/5154387

**Comment by jszwedko** [comment]
Closing since this was just used for tracking.

**Comment by polarathene** [comment]
@jszwedko you may want to update [this docs page](https://vector.dev/docs/about/under-the-hood/architecture/data-model/log/#schemas) which links here to track progress?:

![image](https://github.com/vectordotdev/vector/assets/5098581/b2c4bcb4-c0c7-49ac-be7c-75cced2fdd80)

**Comment by jszwedko** [comment]
> @jszwedko you may want to update [this docs page](https://vector.dev/docs/about/under-the-hood/architecture/data-model/log/#schemas) which links here to track progress?:
> 
> ![image](https://private-user-images.githubusercontent.com/5098581/285635772-b2c4bcb4-c0c7-49ac-be7c-75cced2fdd80.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDEyMDAxODIsIm5iZiI6MTcwMTE5OTg4MiwicGF0aCI6Ii81MDk4NTgxLzI4NTYzNTc3Mi1iMmM0YmNiNC1jMGM3LTQ5YWMtYmU3Yy03NWNjZWQyZmRkODAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQUlXTkpZQVg0Q1NWRUg1M0ElMkYyMDIzMTEyOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyMzExMjhUMTkzMTIyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9M2VmYjE4ODYxYWUyNTI2OTRjMWRmM2JmNjcwZDA5MTIwYmFmODA2ZmQ0Njc2Yjk3ZmIzOTAyZDZmNzFjZjZhZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.bWieup1jI6FDYnrWDIv3P_xZTf8e-RktIqY2VlpK8IA)

Thanks for pointing that out! I opened https://github.com/vectordotdev/vector/pull/19256
